File: WoofIOS/SwiftPackages/Presentation/Navigation/Sources/PresentationNavigation/WoofBackStack.swift
import Combine
import DI
import Foundation

@Single
public final class WoofBackStack: ObservableObject {
    @Published public private(set) var entries: [WoofDestination] = [.home]

    public init() {}

    public func push(_ destination: WoofDestination) {
        entries.append(destination)
    }

    public func pop() {
        if entries.count > 1 {
            entries.removeLast()
        }
    }

    public func popTo(count: Int) {
        entries = Array(entries.prefix(max(count, 1)))
    }
}

File: WoofAndroid/presentation/navigation/src/main/kotlin/com/perrystreet/woof/presentation/navigation/WoofBackStack.kt
package com.perrystreet.woof.presentation.navigation

import androidx.compose.runtime.mutableStateListOf
import org.koin.core.annotation.Single

@Single
class WoofBackStack {
    private val stack = mutableStateListOf<WoofDestination>(WoofDestination.Home)

    val entries: List<WoofDestination>
        get() = stack

    fun push(destination: WoofDestination) {
        stack += destination
    }

    fun pop() {
        if (stack.size > 1) {
            stack.removeAt(stack.lastIndex)
        }
    }
}
